### PR TITLE
Add a byte-backed writer/seeker

### PR DESCRIPTION
**Pull request by dwbuiten**

### CODE REVIEW

File: byteswriter/writer.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+func NewPreallocated(n int) *Writer {
+	ret := new(Writer)
+	ret.buf = make([]byte, 0, n)
+	return ret
+}
+
+// Size returns the current size of the buffer.
+func (w *Writer) Size() int {
+	return len(w.buf)
+}
```
Comment (justinruggles) [inline byteswriter/writer.go]: All the other Size() functions in the io and related packages return int64. Probably best to be consistent with them.

File: byteswriter/writer.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	w.pos = int(off)
+
+	return off, nil
+}
+
+// Write writes to the underlying buffer and increases size as necessary.
+func (w *Writer) Write(buf []byte) (int, error) {
+	if w.pos == len(w.buf) {
+		w.buf = append(w.buf, buf...)
```
Comment (justinruggles) [inline byteswriter/writer.go]: Do you know if the performance of this is comparable to checking capacity, reallocating if needed, and using copy()?

**Comment by justinruggles** [comment]
I think it generally needs some comments to clarify the limitations in terms of things like concurrency and use of the returned slice from Bytes()

File: byteswriter/writer.go (hunk 1 above)
Comment (dwbuiten) [inline byteswriter/writer.go]: Will do.

**Comment by dwbuiten** [comment]
Like most writers, conurrent writing is filed under "don't". I'll add an explicity comment.

As for the return value of `Bytes()`, I assume something like this is what you mean: https://golang.org/pkg/bytes/#Buffer.Bytes

File: byteswriter/writer.go (hunk 2 above)
Comment (dwbuiten) [inline byteswriter/writer.go]: I thought about that too, but I could find no literature on the perf differences, except this: https://gist.github.com/xogeny/b819af6a0cf8ba1caaef which seems to indicate it is about the same?

The main difference, I think, is that `append` will increase the capcaity of the slice as it wants, whereas a `make`/`copy` will have whatever capacity we choose.

File: byteswriter/writer_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	testdata1 := []byte{1, 2, 3, 4, 5, 6, 7, 8}
+	testdata2 := []byte{11, 12, 13, 14}
+	testdata3 := []byte{21, 22, 23, 24}
+
+	w := NewPreallocated(4)
+	n, err := w.Write(testdata1)
+	if err != nil {
+		t.Error(err)
+	} else if n != len(testdata1) {
+		t.Errorf("mismtatched sizes")
```
Comment (justinruggles) [inline byteswriter/writer_test.go]: typo

Comment (justinruggles) [inline byteswriter/writer_test.go]: and several places below

File: byteswriter/writer_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if err == nil {
+		t.Errorf("whence not checked properly")
+	}
+	_, err = w.Seek(-1, io.SeekStart)
+	if err == nil {
+		t.Errorf("offset start not checked properly")
+	}
+	_, err = w.Seek(-1, io.SeekEnd)
+	if err == nil {
+		t.Errorf("offset end not checked porperly")
```
Comment (justinruggles) [inline byteswriter/writer_test.go]: typo

File: byteswriter/writer.go (hunk 2 above)
Comment (justinruggles) [inline byteswriter/writer.go]: looks ok then

File: byteswriter/writer.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	case io.SeekEnd:
+		off = int64(len(w.buf)) - offset
+	default:
+		return 0, fmt.Errorf("invalid whence")
+	}
+
+	if off < 0 {
+		return 0, fmt.Errorf("cannot seek before start of buffer")
+	} else if off > int64(len(w.buf)) {
+		return 0, fmt.Errorf("cannot seek past end of buffer")
```
Comment (justinruggles) [inline byteswriter/writer.go]: According to the io.Seeker documentation, seeking past the end is valid.
```
Seeking to an offset before the start of the file is an error. Seeking to any positive offset is legal, but the behavior of subsequent I/O operations on the underlying object is implementation-dependent.
```

File: byteswriter/writer.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+// Write writes to the underlying buffer and increases size as necessary.
+func (w *Writer) Write(buf []byte) (int, error) {
+	if w.pos == len(w.buf) {
+		w.buf = append(w.buf, buf...)
+	} else if w.pos+len(buf) <= len(w.buf) {
+		copy(w.buf[w.pos:], buf)
+	} else if w.pos+len(buf) > len(w.buf) {
+		overlap := len(w.buf) - w.pos
+		copy(w.buf[w.pos:], buf)
+		w.buf = append(w.buf, buf[overlap:]...)
```
Comment (justinruggles) [inline byteswriter/writer.go]: `copy()` returns the number of elements copied, so you can avoid calculating `overlap` separately by using the return value.

File: byteswriter/writer.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+// for ensuring this.
+package byteswriter
+
+import (
+	"fmt"
+	"io"
+)
+
+// A Writer implements a WriteSeeker interface backed by a
+// synamically expanding buffer.
```
Comment (justinruggles) [inline byteswriter/writer.go]: dynamically?

File: byteswriter/writer.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+type Writer struct {
+	buf []byte
+	pos int
+}
+
+// New returns a new writer with an initial alloction of 4KB.
```
Comment (justinruggles) [inline byteswriter/writer.go]: allocation

**Pull request merged**